sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -3410,32 +3410,60 @@ case class ArrayDistinct(child: Expression)
   }
 
   override def nullSafeEval(array: Any): Any = {
-    val data = array.asInstanceOf[ArrayData].toArray[AnyRef](elementType)
+    val data = array.asInstanceOf[ArrayData]
     doEvaluation(data)
   }
 
   @transient private lazy val doEvaluation = if (TypeUtils.typeWithProperEquals(elementType)) {
-    (data: Array[AnyRef]) => new GenericArrayData(data.distinct.asInstanceOf[Array[Any]])
+    (array: ArrayData) =>
+      val arrayBuffer = new scala.collection.mutable.ArrayBuffer[Any]
+      val hs = new SQLOpenHashSet[Any]()
+      val (isNaN, valueNaN) = SQLOpenHashSet.isNaNFuncAndValueNaN(elementType)
```

Comment:
`isNaN` and `valueNaN` are only used by the next function call `withNaNCheckFunc`, we can calculate them in `withNaNCheckFunc`

Reply:
> `isNaN` and `valueNaN` are only used by the next function call `withNaNCheckFunc`, we can calculate them in `withNaNCheckFunc`

Yea, after  make it return a partial function, the value will be calculated once. Done